README

QTScope - a simple, plugin based oscilloscope program

Copyright (C) 2004 Matthias Hennig <hennig@cn.stir.ac.uk>
Copyright (C) 2005 Bernd Porr <berndporr@f2s.com>
Copyright (C) 2013 Steffen Mauch <steffen.mauch@gmail.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


SUMMARY:
--------

QTScope is a program to display the signals acquired from daq devices
via the comedi library. It is based on the QT Widget Library
and provides a plugin architecture which makes it easy to
write own visualisation plugins.

It was originally written specifically for the USB DUX board
(http://www.linux-usb-daq.co.uk/), but today the aim is to be compatible
with any board with a comedi driver. 
The data aquisition is done in regular
intervals. Under 10000Hz data data is scrolled into the data window.
Above 10000Hz the program does snapshots of 5000 samples and
displays them every second on the screen.


WHAT YOU WILL NEED:
-------------------

This program was developed/tested using a linux distribution
with and QT library 4.x.
the following software should be installed:

Comedi: a free driver/library framework for a number of data
acquisition devices http://www.comedi.org

QT (Version > 4):
http://qt-project.org/

Qwt Widget Library 6.x:
For details about this nice and useful project, see:
http://qwt.sourceforge.net/

For the SHWFS plugin Eigen 3
http://eigen.tuxfamily.org
(for WFS plugin otherwise in
plugins/plugins.pro line with 'wfs \'
has to be removed!)


INSTALLATION:
-------------

1) provide the paths to QT:

   export QTDIR=/path/to/qt

2) qmake qtscope.pro

   (sometimes qmake may not be directly accessible, 
    then you can usually find it under
   /usr/lib/qt4/bin/qmake or /usr/share/qt4/bin/qmake)

3) make

4) As root:

   make install
   
   This will install the program to /usr/local/bin 
   and the plugins to /usr/local/lib/qtscope/plugins.
   
   alternatively
   sudo ./bin/qtscope
   if you just want to try QTScope!
   

TODO:
-----

A lot!

Newer versions will probably contain these planned extensions:

- more visualisation plugins
- data storage facility
- filter plugins for data processing
- plugin for digital input and output (some basics already done)
- plugin for analog output (some basics already done) 


WRITING PLUGINS:
----------------

It is extremely simple to write a plugin for QTScope! A plugin
basically just frequently receives the new data and is free to do
whatever it likes. A QTimer object is responsible for the readout. The
base class for a plugin is called "scopePlotPlugin". Each plugin needs
the following data structure to announce its existence:

pluginInfo myPluginInfo = {...};

whith the tye:

typedef struct pluginInfo{
  QString name;          // the name
  QString type;          // the type, so far not unused (use "Plot")
  int channelsRequest;   // number of channels the plugin requests
  int type_comedi;       // entry of comedi_subdevice_type which the
                         // plugin does support, i.e. COMEDI_SUBD_AI
};

enum comedi_subdevice_type {
  COMEDI_SUBD_UNUSED,     /* unused by driver */
  COMEDI_SUBD_AI,         /* analog input */
  COMEDI_SUBD_AO,         /* analog output */
  COMEDI_SUBD_DI,         /* digital input */
  COMEDI_SUBD_DO,         /* digital output */
  COMEDI_SUBD_DIO,        /* digital input/output */
  COMEDI_SUBD_COUNTER,    /* counter */
  COMEDI_SUBD_TIMER,      /* timer */
  COMEDI_SUBD_MEMORY,     /* memory, EEPROM, DPRAM */
  COMEDI_SUBD_CALIB,      /* calibration DACs */
  COMEDI_SUBD_PROC,       /* processor, DSP */
  COMEDI_SUBD_SERIAL,     /* serial IO */
  COMEDI_SUBD_PWM         /* PWM */
};

The following methods are declared as virtual in the base class:

- virtual void setDataSource(double *) = 0;
  QTScope creates arrays of the type double that contain the data. 
  The plugin just gets the pointer.
  If the plugin expects more than one source, 
  this function is called repeatedly. 

- virtual void insertValues(int) = 0;
  Called when num new data values are available.

- virtual void replot() = 0;
  A screen refresh is due for the plugin.

- samplingRateChanged() = 0;
  The sampling rate "freq" has changed


ChangeLog
Monday 27 January 2014

	1.0 alpha (Steffen Mauch)
	ported to QT4 without needing QT3 suppot
	fixed nasty bugs and added more plugins
	but they are not all ready till now
	improved plugin selection by checking
	available subdevices of comedi

Tuesday 1 February 2005

	Release 0.32 (Bernd Porr)
	The sampling rate can now be changed while the program
	is running. Cleaned up a lot.

Monday 15 November 2004

	Release 0.3 (Bernd Porr)
	The sampling rate is now adjustable. For rates >10kHz, 5000 samples
	are read and thye screen is updated every second.

Wednesday 28 April 2004

	Release 0.2
	Compiles now with gcc >3 (thanks to David Schleef)
	a small change to the plugin interface
	added a x-y plot plugin

Saturday 18 April 2004

	Release 0.1.1
	Fixes and small improvements.
	Qwt library included in the distribution.

Wednesday 7 April 2004

	Initial release 0.1
	ChangeLog Created.


plugins/wfs/wfs.cpp
/***************************************************************************
 *   Copyright (C) 2013 by Steffen Mauch                                   *
 *   steffen.mauch@gmail.com                                               *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include "wfs.h"
// for the data formats
#include <comedilib.h>
#include <QPainter>
#include <iostream>

pluginInfo myPluginInfo = {"WFS Plugin", "Plot", 1, COMEDI_SUBD_MEMORY};

scopePlotPlugin *createPlugin (QTScope* caller, QWidget* parent, const char* name, int id, Qt::WindowFlags wflags,int numberOfSamples) {
	return new wfs(caller, parent, name, id, wflags,numberOfSamples);
}

using std::cout;

wfs::wfs(QTScope* caller, QWidget* parent, const char* name, int id, Qt::WindowFlags wflags, int numberOfSamples)
	: scopePlotPlugin(caller, parent, name, id, wflags, numberOfSamples)
{
	cout << "wfs::wfs: wfs plugin generated\n";
	callingWidget = caller;
	idThis = id;
	
	//set sizePolicy (important for heightForWidth)
	QSizePolicy sizePolicy(QSizePolicy::Minimum, QSizePolicy::Minimum);
	sizePolicy.setHeightForWidth(true);
	setSizePolicy(sizePolicy);
	
	vbox_layout = new QVBoxLayout();
	hbox_layout = new QHBoxLayout();
	
	d_plot = new Plot();
	
	gradient_plot = new gradientPlot( );
	
	en_contour = new QPushButton( tr("enable contour") );
	en_contour->setCheckable( TRUE );
	
	en_nullref = new QPushButton( tr("enable reference") );
	en_nullref->setCheckable( TRUE );
	
	clear_nullref = new QPushButton( tr("clear reference") );
	
	hbox_layout->addWidget(en_nullref);
	hbox_layout->addWidget(clear_nullref);
	
	vbox_layout->addLayout(hbox_layout);

	vbox_layout->addWidget(en_contour);
	vbox_layout->addWidget(gradient_plot);
	vbox_layout->addWidget(d_plot);
	d_plot->hide();

	// connect signals to buttons
	hbox_layout->connect(clear_nullref, SIGNAL( clicked() ),
		this, SLOT( clearReference() ) );
	vbox_layout->connect(en_nullref, SIGNAL( clicked() ),
		this, SLOT( setReference() ) );
	vbox_layout->connect(en_contour, SIGNAL( clicked() ),
		this, SLOT( enableCountour() ) );

    setLayout(vbox_layout);
     
	//setWidget( gradient_plot );
    //setWidget(d_plot);
    
    setWindowTitle("Wavefront Sensor plugin");
    
	show();
	resize(400,400);
	
}

void wfs::enableCountour(){
	qDebug() << "called function: " << __func__ << endl;
	
	if ( en_contour->isChecked() ) {
		d_plot->show();
		gradient_plot->hide();
		//this->resize(400,200);
	} else {
		d_plot->hide();
		gradient_plot->show();
	}
}

void wfs::getComedi(comedi_t *comediDevice1){
	qDebug() << "called function: " << __func__ << endl;
	comediDevice = comediDevice1;
	
	comediSubdevice = comedi_find_subdevice_by_type(comediDevice,COMEDI_SUBD_MEMORY,0);
	
	timer = new QTimer(this);
	connect(timer, SIGNAL(timeout()), this, SLOT(slotReadData()));
	timer->start(500);
	
}

void wfs::slotReadData(){
	qDebug() << "called function: " << __func__ << endl;
	int count = 196;
	lsampl_t data[count];
	double data_x[count];
	double data_y[count];
	
	int retval = 0;
	retval = comedi_data_read_n(comediDevice, comediSubdevice, 0, 0, AREF_GROUND, &data[0], 100);
	
	if(retval < 0)	{
		comedi_perror("comedi_data_read");
		return;
	}
	
	retval = comedi_data_read_n(comediDevice, comediSubdevice, 100, 0, AREF_GROUND, &data[100], 96);
	
	if(retval < 0)	{
		comedi_perror("comedi_data_read");
		return;
	}
	
	for( int k = 0; k < count; k++ ){
		//printf("%d\n", data[k]);
		data_x[k] = double(data[k]);
	}
	
	// enable this for a fake tip/tilt wavefront
	#if 1
	for( int k=0; k<14; k++ ){
		for( int l=0; l<14; l++ ){
			data_x[k*14+l] = 8.0f+k*16+4;
            data_y[k*14+l] = 8.0f+l*16+0;
		}
	}
	#endif
	if ( en_contour->isChecked() ) 
		d_plot->setData( &data_x[0], &data_y[0] );
	else
		gradient_plot->setData( &data_x[0], &data_y[0] );
}

void wfs::clearReference(){
	qDebug() << "called function: " << __func__ << endl;
}

void wfs::setReference(){
	qDebug() << "called function: " << __func__ << endl;
}

QSize wfs::sizeHint() const
{
	return QSize(480, 550);
	//cout << "getting called\n";
}

wfs::~wfs()
{
}


